File: 20. Valid Parentheses.cpp
class Solution {
public:
    // stack: O(N), 0.29%, ok
    bool isValid(string s) {
        if(s.size() % 2) return false;
        stack<char> cache;
        unordered_map<char, char> p = {{']', '['}, {')', '('}, {'}', '{'}};
        for(auto c: s){
            if(cache.empty() || c == '[' || c == '(' || c == '{')
                cache.push(c);
            else{
                if(cache.empty() || p[c] != cache.top()) return false;
                cache.pop();
            }
        }
        return cache.empty();
    }
};

File: 20. Valid Parentheses.py
class Solution(object):
    
    # stack: O(N) time & space
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack, paren = [], {")":"(", "}":"{", "]":"["}
        for c in s:
            if c in "({[":
                stack.append(c)
            else:
                if stack == [] or stack.pop() != paren[c]:   # bug: stack is empty
                    return False
        return len(stack) == 0
                
        